packages/ui/src/components/Visualization/Custom/hooks/paste-step.hook.tsx

```diff
@@ -3,54 +3,75 @@ import { AddStepMode, IVisualizationNode } from '../../../../models/visualizatio
+  const checkClipboardCompatibility = useCallback(
+    (pastedNodeValue: IClipboardCopyObject | null): boolean => {
+      if (!pastedNodeValue) return false;
 
-        /** Validate the pasted node */
-        if (isSameType) {
-          /** Get compatible nodes and the location where can be introduced */
-          const filter = entitiesContext.camelResource.getCompatibleComponents(
-            mode,
-            vizNode.data,
-            vizNode.getComponentSchema()?.definition,
-          );
+      const pastedNodeType = pastedNodeValue.type;
+      const baseNodeType = entitiesContext.camelResource.getType();
+      const isSameType = pastedNodeType === baseNodeType;
 
-          /** Check paste compatibility */
-          const compatibility = catalogModalContext?.checkCompatibility(pastedNodeValue.name, filter) ?? false;
+      /** Validate the pasted node */
+      if (!isSameType) return false;
+      /** Get compatible nodes and the location where can be introduced */
+      const filter = entitiesContext.camelResource.getCompatibleComponents(
+        mode,
+        vizNode.data,
+        vizNode.getComponentSchema()?.definition,
+      );
 
-          setIsCompatible(compatibility);
-          return;
-        }
-      }
+      /** Check paste compatibility */
+      return catalogModalContext?.checkCompatibility(pastedNodeValue.name, filter) ?? false;
+    },
+    [catalogModalContext, entitiesContext, mode, vizNode],
+  );
 
-      setIsCompatible(false);
+  /** Compatibility check on effect */
+  useEffect(() => {
+    const validate = () => {
+      navigator.permissions
+        .query({ name: permissionName })
+        .then(async () => {
+          const pastedNodeValue = await ClipboardManager.paste();
+          const compatible = checkClipboardCompatibility(pastedNodeValue);
+          setIsCompatible(compatible);
+        })
+        .catch(() => {
+          // fallback to allow pasting incase of permission issues (for Firefox or other browsers)
+          setIsCompatible(true);
+        });
     };
 
-    checkCompatibility();
-  }, [entitiesContext, catalogModalContext, mode, vizNode]);
+    validate();
+  }, [checkClipboardCompatibility]);
 
   const onPasteStep = useCallback(async () => {
     if (!vizNode || !entitiesContext) return;
 
     const pastedNodeValue = await ClipboardManager.paste();
-    if (pastedNodeValue) {
-      /** Paste copied node to the entities */
-      vizNode.pasteBaseEntityStep(pastedNodeValue, mode);
-      /** Update entity */
-      entitiesContext.updateEntitiesFromCamelResource();
-    }
-  }, [entitiesContext, mode, vizNode]);
+    if (!pastedNodeValue) return;
+
+    navigator.permissions.query({ name: permissionName }).catch(() => {
+      const compatible = checkClipboardCompatibility(pastedNodeValue);
+      if (!compatible) {
+        alert('Pasted node is not compatible with the current context.');
+        return;
+      }
+    });
```

Comment:
A few comments around this function:

1. `.catch` will resolve after `vizNode.pasteBaseEntityStep` is called.
2. The return won't help because that return is for the catch function; it's not working for the `onPasteStep` function
3. Is it necessary to perform the clipboard check again? I think if it failed the first time, it would be fail the second one as well, so maybe just check if the steps are compatible, otherwise we get out of the function.